File: examples/python/rocket_ipopt.py
# -*- coding: utf-8 -*-
from casadi import *
from numpy import *
import matplotlib.pyplot as plt

# Construct an explicit euler integrator
def create_integrator_euler():
  u = SX("u") # control for one segment

  # Initial position
  s0 = SX("s0") # initial position
  v0 = SX("v0") # initial speed
  m0 = SX("m0") # initial mass

  t0 = SX("t0") # initial time
  tf = SX ("tf") # final time

  nj = 1000 # Number of integration steps per control segment
  dt = (tf-t0)/nj # time step
  alpha = SX(0.05)    # friction
  beta = SX(0.1)      # fuel consumption rate

  # Integrate over the interval with Euler forward
  s = s0;   v = v0;    m = m0

  dm = -dt*beta*u*u
  for j in range(nj):
    s += dt*v
    v += dt * (u-alpha*v*v)/m
    m += dm

  # State vector
  x =  [s,v,m]
  x0 = [s0,v0,m0];

  # Input to the integrator function being created
  integrator_in = INTEGRATOR_NUM_IN * [[]]
  integrator_in[INTEGRATOR_T0] = [t0]
  integrator_in[INTEGRATOR_TF] = [tf]
  integrator_in[INTEGRATOR_X0] =  x0
  integrator_in[INTEGRATOR_P]  =  [u]

  # Create a dummy state derivative vector
  xp0 = create_symbolic("x",len(x))
  integrator_in[INTEGRATOR_XP0] = xp0

  # Create the explicit Euler integrator
  integrator = SXFunction(integrator_in,[x])
  return integrator

# Construct an explicit multistep integrator from Sundials
def create_integrator_cvodes():
  # Time 
  t = SX("t")

  # Differential states
  s = SX("s"); v = SX("v"); m = SX("m")
  y = [s,v,m]

  # Control
  u = SX("u")

  alpha = 0.05 # friction
  beta = 0.1 # fuel consumption rate
  
  # Differential equation
  sdot = v
  vdot = (u-alpha*v*v)/m
  mdot = -beta*u*u
  rhs = [sdot,vdot,mdot]

  # Initial conditions
  y0 = [0,0,1]
  
  # Input of the ode rhs
  ffcn_in = ODE_NUM_IN * [[]]
  ffcn_in[ODE_T] = [t]
  ffcn_in[ODE_Y] =  y
  ffcn_in[ODE_P] = [u]
  
  # ODE right hand side
  ffcn = SXFunction(ffcn_in,[rhs])
  ffcn.setOption("name","ODE right hand side")
  ffcn.setOption("ad_order",1)

  # Explicit integrator (CVODES)
  integrator = CVodesIntegrator(ffcn)
  
  # Set options
  #integrator.setOption("exact_jacobian",True)
  #integrator.setOption("linear_multistep_method","bdf") # adams or bdf
  #integrator.setOption("nonlinear_solver_iteration","newton") # newton or functional
  integrator.setOption("ad_order",1)
  integrator.setOption("fsens_err_con",True)
  integrator.setOption("quad_err_con",True)
  integrator.setOption("abstol",1e-6)
  integrator.setOption("reltol",1e-6)
  #integrator.setOption("fsens_all_at_once",False)

  return integrator



# Main function

# Time length
T = 10.0

# Shooting length
nu = 1000 # Number of control segments
DT = double(T)/nu

# Initial position, speed and mass
s0 = 0 # initial position
v0 = 0 # initial speed
m0 = 1 # initial mass
X0 = [s0,v0,m0]

# Create integrators
integrator_euler = create_integrator_euler()
integrator_cvodes = create_integrator_cvodes()

for integrator in [integrator_euler, integrator_cvodes]:
  # Enable AD
  integrator.setOption("ad_order",1)

  # Initialize the integrator
  integrator.init()

  # control for all segments
  U = MX("U",nu)

  # Dummy input corresponding to the state derivative
  xdot = MX([0,0,0])

  # Integrate over all intervals
  X=MX(X0)
  for k in range(nu):
    X = integrator([MX(k*DT),MX((k+1)*DT),X,U[k],xdot])  # build up a graph with function calls

  # Objective function
  F = inner_prod(U,U)

  # Terminal constraints
  G = vertcat(X[0],X[1])
  
  # Create the NLP
  ffcn = MXFunction([U],[F]) # objective function
  gfcn = MXFunction([U],[G]) # constraint function

  # Allocate an NLP solver
  #solver = LiftedNewtonSolver(ffcn,gfcn)
  solver = IpoptSolver(ffcn,gfcn)
  
  # Set options
  solver.setOption("tol",1e-10)
  solver.setOption("hessian_approximation","limited-memory");

  # initialize the solver
  solver.init()

  # Bounds on u and initial condition
  Umin = nu * [-10] # lower bound
  solver.setInput(Umin,NLP_LBX)

  Umax = nu * [10]  # upper bound
  solver.setInput(Umax,NLP_UBX)

  Usol = nu * [0.4] # initial guess
  solver.setInput(Usol,NLP_X_INIT)

  # Bounds on g
  Gmin = Gmax = [10, 0]
  solver.setInput(Gmin,NLP_LBG)
  solver.setInput(Gmax,NLP_UBG)

  # Solve the problem
  solver.solve()

  # Get the solution
  xopt = solver.getOutput(NLP_X_OPT)

  # Plot the optimal trajectory
  plt.figure()
  plt.plot(xopt) 
  plt.show()
  
  

File: examples/cplusplus/rocket_ipopt.cpp
/*
 *    This file is part of CasADi.
 *
 *    CasADi -- A symbolic framework for dynamic optimization.
 *    Copyright (C) 2010 by Joel Andersson, Moritz Diehl, K.U.Leuven. All rights reserved.
 *
 *    CasADi is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 3 of the License, or (at your option) any later version.
 *
 *    CasADi is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with CasADi; if not, write to the Free Software
 *    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <iostream>
#include <fstream>
#include <ctime>
#include <symbolic/casadi.hpp>
#include <interfaces/ipopt/ipopt_solver.hpp>
#include <symbolic/stl_vector_tools.hpp>

using namespace CasADi;
using namespace std;

int main(){
    
  cout << "program started" << endl;
      
  // Dimensions
  int nu = 20;  // Number of control segments
  int nj = 100; // Number of integration steps per control segment

  // optimization variable
  vector<SX> u = ssym("u",nu).data(); // control

  SX s_0 = 0; // initial position
  SX v_0 = 0; // initial speed
  SX m_0 = 1; // initial mass
  
  SX dt = 10.0/(nj*nu); // time step
  SX alpha = 0.05; // friction
  SX beta = 0.1; // fuel consumption rate

  // Trajectory
  vector<SX> s_traj(nu), v_traj(nu), m_traj(nu);

  // Integrate over the interval with Euler forward
  SX s = s_0, v = v_0, m = m_0;
  for(int k=0; k<nu; ++k){
    for(int j=0; j<nj; ++j){
      s += dt*v;
      v += dt / m * (u[k]- alpha * v*v);
      m += -dt * beta*u[k]*u[k];
    }
    s_traj[k] = s;
    v_traj[k] = v;
    m_traj[k] = m;
  }

  // Objective function
  SX f = 0;
  for(int i=0; i<u.size(); ++i)
    f += u[i]*u[i];
    
  // Terminal constraints
  vector<SX> g(2);
  g[0] = s;
  g[1] = v;
  g.insert(g.end(),v_traj.begin(),v_traj.end());
  
  // Create the NLP
  SXFunction ffcn(u,f); // objective function
  SXFunction gfcn(u,g); // constraint
  
  // Allocate an NLP solver
  IpoptSolver solver(ffcn,gfcn);

  // initialize the solver
  solver.init();

  // Bounds on u and initial condition
  vector<double> umin(nu), umax(nu), uinit(nu);
  for(int i=0; i<nu; ++i){
    umin[i] = -10;
    umax[i] =  10;
    uinit[i] = 0.4;
  }
  solver.setInput(umin,"lbx");
  solver.setInput(umax,"ubx");
  solver.setInput(uinit,"x0");
  
  // Bounds on g
  vector<double> gmin(2), gmax(2);
  gmin[0] = gmax[0] = 10;
  gmin[1] = gmax[1] =  0;
  gmin.resize(2+nu, -numeric_limits<double>::infinity());
  gmax.resize(2+nu, 1.1);
  
  solver.setInput(gmin,"lbg");
  solver.setInput(gmax,"ubg");

  // Solve the problem
  solver.solve();
  
  // Print the optimal cost
  double cost;
  solver.getOutput(cost,"f");
  cout << "optimal cost: " << cost << endl;

  // Print the optimal solution
  vector<double> uopt(nu);
  solver.getOutput(uopt,"x");
  cout << "optimal control: " << uopt << endl;

  // Get the state trajectory
  vector<double> sopt(nu), vopt(nu), mopt(nu);
  vector<Matrix<SX> > xfcn_out(3);
  xfcn_out[0] = s_traj;
  xfcn_out[1] = v_traj;
  xfcn_out[2] = m_traj;
  SXFunction xfcn(u,xfcn_out);
  xfcn.init();
  xfcn.setInput(uopt);
  xfcn.evaluate();
  xfcn.getOutput(sopt,0);
  xfcn.getOutput(vopt,1);
  xfcn.getOutput(mopt,2);
  cout << "position: " << sopt << endl;
  cout << "velocity: " << vopt << endl;
  cout << "mass:     " << mopt << endl;

  // Create Matlab script to plot the solution
  ofstream file;
  string filename = "rocket_ipopt_results.m";
  file.open(filename.c_str());
  file << "% Results file from " __FILE__ << endl;
  file << "% Generated " __DATE__ " at " __TIME__ << endl;
  file << endl;
  file << "cost = " << cost << ";" << endl;
  file << "u = " << uopt << ";" << endl;

  // Save results to file
  file << "t = linspace(0,10.0," << nu << ");"<< endl;
  file << "s = " << sopt << ";" << endl;
  file << "v = " << vopt << ";" << endl;
  file << "m = " << mopt << ";" << endl;
  
  // Finalize the results file
  file << endl;
  file << "% Plot the results" << endl;
  file << "figure(1);" << endl;
  file << "clf;" << endl << endl;
  
  file << "subplot(2,2,1);" << endl;
  file << "plot(t,s);" << endl;
  file << "grid on;" << endl;
  file << "xlabel('time [s]');" << endl;
  file << "ylabel('position [m]');" << endl << endl;
  
  file << "subplot(2,2,2);" << endl;
  file << "plot(t,v);" << endl;
  file << "grid on;" << endl;
  file << "xlabel('time [s]');" << endl;
  file << "ylabel('velocity [m/s]');" << endl << endl;
  
  file << "subplot(2,2,3);" << endl;
  file << "plot(t,m);" << endl;
  file << "grid on;" << endl;
  file << "xlabel('time [s]');" << endl;
  file << "ylabel('mass [kg]');" << endl << endl;
  
  file << "subplot(2,2,4);" << endl;
  file << "plot(t,u);" << endl;
  file << "grid on;" << endl;
  file << "xlabel('time [s]');" << endl;
  file << "ylabel('Thrust [kg m/s^2]');" << endl << endl;
  
  file.close();
  cout << "Results saved to \"" << filename << "\"" << endl;

  return 0;
}

